### DISCUSSION TITLE
How to evaluate if a entity "insert" are inserted in a PaperSpace or in a Model Space

### DISCUSSION THREAD

**Discussion by nquinterno** [mozman/ezdxf#680] [Q&A]
Hi, 

I need to evaluate if a entity "insert" is in modelspace or in a paperspace, but all the ways that I tryed dont give me positive results.

Someone can help me?

![consulta_hithub](https://user-images.githubusercontent.com/22855208/167314713-267bbaf6-d333-40ac-beeb-e619c9d0153e.png)

**Comment by mozman** [comment]
The `paperspace` attribute is an old relict and it is not a reliable source for the parent layout. This attribute is for newer DXF versions not necessary because each entity has an owner tag which points to the parent layout and each entity can only exist in one layout. 

There are more kind of layouts than `modelspace` and `paperspace`. For your use case you also have to consider the `block` layout and for more advanced usage is also the `virtual state` of entities important where an entity is not assigned to any layout.


```Python

from ezdxf.entities import is_graphic_entity

...

for e in doc.query("*"):
    if not is_graphic_entity(e):
        continue
    layout = e.get_layout()
    if layout is not None:
        if layout.is_modelspace:
            print(f"entity {e} is located in the modelspace")
        elif layout.is_any_paperspace:
            print(f"entity {e} is located in paperspace '{layout.name}'")
        elif layout.is_block_layout:
            print(f"entity {e} is located in block '{layout.name}'")
    else:
        print(f"entity {e} has no parent layout")
```

Be aware, if you query the whole document you can get entities that are not subclasses of `DXFGraphic` which do not have a parent layout 
 and therefore do not support the `get_layout()` method.

**Reply by nquinterno** [reply]
I understand that exist more layouts than modelspace or paper space in the DXF Structure and this is my problen, the entities "insert" are in a Block Layout and when i evaluate the layout asociated to the insert the code return this Block Layout, but i am interesting to know if entity "insert" was inserted in the model space o any paperspace for the drawing, because this there are 2 only options thas the CAD aplicación user  have (insert the entity in the model space or in the any paper space). If an block was inserted in the model space of the dxf drawing i need to capture with an alert flag

So, if i understand correctly, with de "owner" tag for an entity "insert" can i know if the entity was inserted in model or paper space, rigth?

**Reply by mozman** [reply]
The `INSERT` entity can also be used in a `BLOCK`. If you just wanna know which `INSERT` entities are located in the modelspace query only the modelspace instead of the whole entity database:

```Python
msp = doc.modelspace()
for insert in msp.query("INSERT"):
    raise TypeError(f"entity {e} not allowed in modelspace")

```

**Reply by nquinterno** [reply]
thanks, i will try that
